### PR TITLE
wasm: improve cache policy

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fixes: #42154

this PR depends on PRs https://github.com/envoyproxy/envoy/pull/24384 and https://github.com/istio/istio/pull/42365, so it cannot be backported

### CODE REVIEW

File: istioctl/cmd/authz.go [hunk 1]
```diff
@@ -23,9 +23,9 @@ import (
 	"github.com/spf13/cobra"
 
 	"istio.io/istio/istioctl/pkg/authz"
-	"istio.io/istio/istioctl/pkg/util/configdump"
 	"istio.io/istio/istioctl/pkg/util/handlers"
 	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/util/configdump"
```
Comment (hzxuzhonghu) [inline istioctl/cmd/authz.go:28]: This can be separated, i donot see the correlation with wasm cache

Comment (zirain) [reply istioctl/cmd/authz.go:28]: see module_getter.go

File: pkg/wasm/module_getter.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func NewLocalModuleGetter(adminPort int32, localhostAddr string) ModuleGetter {
+	return &LocalModuleGetter{
+		AdminPort:     adminPort,
+		LocalHostAddr: localhostAddr,
+	}
+}
+
+func (g *LocalModuleGetter) ListUsingModules() (sets.String, error) {
```
Comment (hzxuzhonghu) [inline pkg/wasm/module_getter.go:47]: I am concerned about the perf

**Review by hzxuzhonghu — COMMENTED** [review]
IMO we can solve the issue in this way

Each time we get a wasm, we should update the timestamp simultaneously. 

Since pilot-agent will disconnect with istiod periodically, we can rely on the full xds push each time. 

The caveat is `MaxServerConnectionAge` can be set to super long

File: pkg/wasm/module_getter.go (hunk 2 above)
Comment (howardjohn) [reply pkg/wasm/module_getter.go:47]: +1, I am pretty strongly against using config_dump in anything other than debugging context.

*if* we did, there is a way to filter it though: https://www.envoyproxy.io/docs/envoy/latest/operations/admin#get--config_dump?mask=

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pkg/wasm/module_getter.go (hunk 2 above)
Comment (zirain) [reply pkg/wasm/module_getter.go:47]: I agree with that `config_dump is not an elegant way`, and we need more experiments before introducing `TTL`

Speaking of performance issues, interval should not be very short in a real world, but as your said with `mask` will be better.

**Comment by hzxuzhonghu** [comment]
ECDS requests will include all the resource names that a envoy subscribes, so I think we can hijack all the ecds requests.

**Comment by ingwonsong** [comment]
@zirain 

IMHO, one of easiest way to avoid this issue is to purge the wasm binaries when new ECDS comes, instead of triggering it by Timer.
This should work because ECDS pushes include all the resources as @hzxuzhonghu mentioned.

WDYT?

**Comment by ingwonsong** [comment]
In addition, after re-thinking, I concluded that TTL is not a great choice for this purpose. 
It is because, if Wasm Plugin can be expired in Envoy due to TTL and, at the same time, Istiod is not reachable somehow, it may not work properly.

**Comment by ingwonsong** [comment]
FYI, Experimental PR.

https://github.com/istio/istio/compare/master...ingwonsong:istio:purge-wasm-on-get?expand=1

**Comment by zirain** [comment]
@ingwonsong would be like to send PR for https://github.com/istio/istio/pull/42486#issuecomment-1363405938

**Comment by zirain** [comment]
we should try use ECDS trigger in first stage, then introduce TTL as experimental feature

**Pull request closed without merging**